test/test_grid.py

```diff
@@ -534,9 +509,225 @@ def test_edge_nodes_euler(self):
+            face_nodes_connectivity = tgrid.ds["Mesh2_face_nodes"].values
+
+            # Create a dictionary to store the face indices for each edge
+            face_nodes_dict = {}
+
+            # Loop through each face and edge to build the dictionary
+            for face_idx, face_edges in enumerate(face_edges_connectivity):
+                for edge_idx in face_edges:
+                    if edge_idx != ux.INT_FILL_VALUE or np.isnan(edge_idx):
+                        edge = edge_nodes_connectivity[edge_idx]
+                        if face_idx not in face_nodes_dict:
+                            face_nodes_dict[face_idx] = []
+                        face_nodes_dict[face_idx].append(edge[0])
+                        face_nodes_dict[face_idx].append(edge[1])
+
+            # Make sure the face_nodes_dict is in the counter-clockwise order and remove duplicate nodes
+            for face_idx, face_nodes in face_nodes_dict.items():
+                # First need to re-position the first two nodes position according to the original face_nodes_connectivity
+                first_edge_correct = np.array([
+                    face_nodes_connectivity[face_idx][0],
+                    face_nodes_connectivity[face_idx][1]
+                ])
+                first_edge = np.array([face_nodes[0], face_nodes[1]])
+
+                first_edge_correct_copy = first_edge_correct.copy()
+                first_edge_copy = first_edge.copy()
+                self.assertTrue(
+                    np.array_equal(first_edge_correct_copy.sort(),
+                                   first_edge_copy.sort()))
+                face_nodes[0] = first_edge_correct[0]
+                face_nodes[1] = first_edge_correct[1]
+                i = 2
+                while i < len(face_nodes):
+                    if face_nodes[i] != face_nodes[i - 1]:
+                        # swap the order
+                        old = face_nodes[i]
+                        face_nodes[i] = face_nodes[i - 1]
+                        face_nodes[i + 1] = old
+                    i += 2
+
+                after_swapped = face_nodes
+
+                after_swapped_remove = [after_swapped[0]]
+
+                for i in range(1, len(after_swapped) - 1):
+                    if after_swapped[i] != after_swapped[i - 1]:
+                        after_swapped_remove.append(after_swapped[i])
+
+                face_nodes_dict[face_idx] = after_swapped_remove
+
+            # Convert the dictionary to a list
+            res_face_nodes_connectivity = []
+            for face_idx in range(len(tgrid.ds["Mesh2_face_edges"].values)):
+                res_face_nodes_connectivity.append(face_nodes_dict[face_idx])
+                while len(res_face_nodes_connectivity[face_idx]
+                         ) < tgrid.ds["Mesh2_face_nodes"].values.shape[1]:
+                    res_face_nodes_connectivity[face_idx].append(
+                        ux.INT_FILL_VALUE)
```

Comment:
Can you give me brief rundown of the motivation for writing this? Also, there is a lot of duplicated code in the next case case. We should consider abstracting these code away into a test helper function that can be called.

Reply:
As explained in the comments, this testcase will revert the `edge_nodes_connectivity` and `face_edges_connectivity` back to the `face_nodes_connectivity`, and will compare the reverted one and the original one and see if they're the same

I can move it to elsewhere. Although I didn't see its future usage in other functions

Reply:
Makes sense, this is a great test.

We wouldn't move it into our API, we would still keep it contained within our tests. It would make understanding the test much easier and would reduce the need to duplicate this section twice for both of your tests.